File: Data Structures & Algorithms/longest-increasing-subsequence/submission-8.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        N = len(nums)
        dp = {}
        def dfs(i,j):
            if i ==N:
                return 0
            if (i,j) in dp:
                return dp[(i,j)]
            LIS = dfs(i+1,j)
            if j == -1 or nums[i]>nums[j]:
                LIS =  max(1+dfs(i+1,i),LIS)
            dp[(i,j)] = LIS
            return LIS
        return dfs(0,-1)





    #     9   1   4   2   3   3   7

    # -1001.  1   4   
    #     1   1   2

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-8.java
class Solution {
    boolean dp[][];
    int sum;
    public boolean canPartition(int[] nums) {
        sum = Arrays.stream(nums).sum();
        if(sum%2!=0){
            return false;
        }
        dp = new boolean[nums.length][sum];
        return dfs(0,0,nums);

        
    }

    public boolean dfs(int i,int s, int[] nums){
        if(s==sum/2){
            return true;
        }
        if(s>sum/2 || i == nums.length){
            return false;
        }
        System.out.println(i+" "+s+" "+sum);
        if(dp[i][s]){
            return dp[i][s];
        }
        dp[i][s] = dfs(i+1,s+nums[i],nums) || dfs(i+1,s,nums) ;
        return dp[i][s];
    }
}
